### PR TITLE
[drop tracking] Visit break expressions 

### LINKED ISSUE #102383: Drop tracking ICE involving await in left hand side of assignment (`-Zdrop-tracking`)
This was found during the most recent crater run for #97334. The test case below is reduced from https://github.com/haydnv/tinychain.

```rust
pub struct File<B> {
    block: B,
}

pub async fn commit<B: Clone>(this: &mut File<B>) {
    async {}.await;
    async {}.await;
    async {}.await;
    async {}.await;

    let file = async { &this }.await;
    *async { &mut this.block }.await = file.block.clone();
}
```

This ICEs with the following message:
```
error: internal compiler error: compiler/rustc_mir_transform/src/generator.rs:760:13: Broken MIR: generator contains type B in MIR, but typeck only knows about {ResumeTy, &mut File<B>, impl Future<Output = ()>, (), impl Future<Output = ()>, impl Future<Output = ()>, impl Future<Output = ()>, impl Future<Output = &&mut File<B>>, &&mut File<B>, File<B>, impl Future<Output = &mut B>} and [&mut File<B>]
  --> src/lib.rs:5:51
   |
5  |   pub async fn commit<B: Clone>(this: &mut File<B>) {
   |  ___________________________________________________^
6  | |     async {}.await;
7  | |     async {}.await;
8  | |     async {}.await;
...  |
12 | |     *async { &mut this.block }.await = file.block.clone();
13 | | }
   | |_^
```

My guess is that the `B` that MIR and typeck disagree on is the one produced by `file.block.clone()`.

Weirdly, if any of the `async {}.await` lines are removed, the program compiles successfully.

### PR DESCRIPTION
**Pull request by eholk**
This fixes https://github.com/rust-lang/rust/issues/102383 by remembering to visit the expression in `break expr` when building the drop tracking CFG. Missing this step was causing an off-by-one error which meant after a number of awaits we'd be
looking for dropped values at the wrong point in the code.

Additionally, this changes the order of traversal for assignment expressions to visit the rhs and then the lhs. This matches what is done elsewhere.

Finally, this improves some of the debugging output (for example, the CFG visualizer) to make it easier to figure out these sorts of issues.

### CODE REVIEW

**Review by vincenzopalazzo — APPROVED** [review]

**Review by albertlarsan68 — CHANGES_REQUESTED** [review]
Tests have moved from `src/test` to `tests`, please [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing) on top of new `master`.
@rustbot author

**Comment by eholk** [comment]
> Tests have moved from `src/test` to `tests`, please [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing) on top of new `master`. @rustbot author

Nice to see that change happen! I updated this PR.

File: compiler/rustc_hir_typeck/src/generator_interior/drop_ranges/cfg_build.rs [hunk 1]
```diff
@@ -304,8 +304,8 @@ impl<'a, 'tcx> Visitor<'tcx> for DropRangeVisitor<'a, 'tcx> {
         let mut reinit = None;
         match expr.kind {
             ExprKind::Assign(lhs, rhs, _) => {
-                self.visit_expr(lhs);
                 self.visit_expr(rhs);
+                self.visit_expr(lhs);
```
Comment (wesleywiser) [inline compiler/rustc_hir_typeck/src/generator_interior/drop_ranges/cfg_build.rs:308]: This matches the behavior described in the Rust Reference for assignment expressions (https://doc.rust-lang.org/reference/expressions/operator-expr.html#basic-assignments) 👍

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @eholk; this looks good to me!

@bors r+ rollup

**Pull request merged**